src/fhirtypes/common.ts

```diff
@@ -58,6 +58,91 @@ export function setPropertyOnDefinitionInstance(
   setPropertyOnInstance(instance, pathParts, assignedValue, fisher);
 }
 
+export function createUsefulSlices(
+  instanceDef: StructureDefinition | ElementDefinition | InstanceDefinition,
+  instanceOfStructureDefinition: StructureDefinition,
+  ruleMap: Map<string, { pathParts: PathPart[] }>,
+  fisher: Fishable
+) {
+  ruleMap.forEach(({ pathParts }, path) => {
+    const nonNumericPath = path.replace(/\[[-+]?\d+\]/g, '');
+    const element = instanceOfStructureDefinition.findElementByPath(nonNumericPath, fisher);
+    if (element) {
+      // go through the parts, and make sure that we have a useful index, and maybe a named slice
+      let current: any = instanceDef;
+      let currentPath = '';
+      for (const [i, pathPart] of pathParts.entries()) {
+        currentPath += `${currentPath ? '.' : ''}${pathPart.base}`;
+        const currentElement = instanceOfStructureDefinition.findElementByPath(currentPath, fisher);
+        const key =
+          pathPart.primitive && i < pathParts.length - 1 ? `_${pathPart.base}` : pathPart.base;
```

Comment:
A comment might be helpful here.  I had to think about this a minute before I had it straight.  Basically, I think this is what is happening?
> // If this is a primitive and the path continues to a nested element of the primitive,
> // then we need to look at the special property that starts with _ instead.